File: solution/2040/main.py
def main():
    input = open(0, "rb").readline
    N = int(input())
    for _ in range(3):
        min_diff = [0]
        prefix_sum = [0]
        prefix = 0
        for ai in map(int, input().split()):
            prefix += ai
            optimal = 1 << 30
            for d, s in zip(min_diff, prefix_sum):
                optimal = min(optimal, prefix - s - d)
            min_diff.append(optimal)
            prefix_sum.append(prefix)
        if min_diff[-1] < 0:
            print("A")
        elif min_diff[-1] > 0:
            print("B")
        else:
            print("D")


main()

File: solution/2040/main.rs
fn main() {
    use std::io::*;
    let stdin = read_to_string(stdin().lock()).unwrap();
    let mut tokens = stdin.split_whitespace();
    let mut next = || tokens.next().unwrap();
    let n = next().parse::<usize>().unwrap();
    for _ in 0..3 {
        let mut min_diff = vec![0];
        let mut prefix_sum = vec![0];
        let mut prefix = 0;
        for _ in 0..n {
            let ai = next().parse::<i32>().unwrap();
            prefix += ai;
            let mut optimal = i32::MAX;
            for (&d, &s) in min_diff.iter().zip(prefix_sum.iter()) {
                optimal = optimal.min(prefix - s - d);
            }
            min_diff.push(optimal);
            prefix_sum.push(prefix);
        }
        let winner = match *min_diff.last().unwrap() {
            ..=-1 => 'A',
            1.. => 'B',
            0 => 'D',
        };
        println!("{winner}");
    }
}
